## Week 10 and 11 Assignment - DATASCI200 Introduction to Data Science Programming, UC Berkeley MIDS

Write code in this Jupyter Notebook to solve the following problems. Please upload this **Notebook** with your solutions to your GitHub repository and provide a link in the last question in gradescope. 

Assignment due date: 11:59PM PT the night before the Week 12 Live Session. Do **NOT** push/upload the data file. 

## Objectives

- Explore and glean insights from a real dataset using pandas
- Practice using pandas for exploratory analysis, information gathering, and discovery
- Practice cleaning data and answering questions

## General Guidelines:

- This is a **real** dataset and so it may contain errors and other pecularities to work through
- This dataset is ~218mb, which will take some time to load (and probably won't load in Google Sheets or Excel)
- If you make assumptions, annotate them in your responses
- While there is one code/markdown cell positioned after each question as a placeholder, some of your code/responses may require multiple cells
- Double-click the markdown cells that say for example **1a answer here:** to enter your written answers. If you need more cells for your written answers, make them markdown cells (rather than code cells)
- This homework assignment is not autograded because of the variety of responses one could give. 
  - Please upload this notebook to the autograder page and the TAs will manually grade it. 
  - Ensure that each cell is run and outputs your answer for ease of grading! 
  - Highly suggest to do a `restart & run all` before uploading your code to ensure everything runs and outputs correctly.
  - Answers without code (or code that runs) will be given 0 points.
- **This is meant to simulate real world data so you will have to do some external research to determine what some of the answers are!** 

## Dataset

You are to analyze campaign contributions to the 2016 U.S. presidential primary races made in California. Use the csv file located here: https://drive.google.com/file/d/1Lgg-PwXQ6TQLDowd6XyBxZw5g1NGWPjB/view?usp=sharing. You should download and save this file in the same folder as this notebook is stored.  This file originally came from the U.S. Federal Election Commission (https://www.fec.gov/).

**DO NOT PUSH THIS FILE TO YOUR GITHUB REPO!**

- Best practice is to not have DATA files in your code repo. As shown below, the default load is outside of the folder this notebook is in. If you change the folder where the file is stored please update the first cell!
- If you do accidentally push the file to your github repo - follow the directions here to fix it: https://docs.google.com/document/d/15Irgb5V5G7pKPWgAerH7FPMpKeQRunbNflaW-hR2hTA/edit?usp=sharing

Documentation for this data can be found here: https://drive.google.com/file/d/11o_SByceenv0NgNMstM-dxC1jL7I9fHL/view?usp=sharing

## Data Questions

You are working for a California state-wide election campaign. Your boss wants you to examine historic 2016 election contribution data to see what zipcodes are more supportive of fundraising for your candidate. 

Your boss asks you to filter out some of the records:
- Only use primary 2016 contribution data (more like how your race is).
- Concentrate on Bernie Sanders as a candidate (most a like your candidate)

The questions your boss wants answered is:
- Which zipcode (5-digit zipcode) had the highest count of contributions and the most dollar amount?
- What day(s) of the month do most people donate?

## Setup

Run the cell below as it will load the data into a pandas dataframe named `contrib`. Note that a custom date parser is defined to speed up loading. If Python were to guess the date format, it would take even longer to load.

In [ ]:
import pandas as pd
import numpy as np
from datetime import datetime

# These commands below set some options for pandas and to have matplotlib show the charts in the notebook
pd.set_option('display.max_rows', 1000)
pd.options.display.float_format = '{:,.2f}'.format

# Define a date parser to pass to read_csv
d = lambda x: datetime.strptime(x, '%d-%b-%y')

# Load the data
# We have this defaulted to the folder OUTSIDE of your repo - please change it as needed
contrib = pd.read_csv("/mnt/c/Users/davidilitzky.REDMOND/Berkeley/P00000001-CA.csv", index_col=False, parse_dates=['contb_receipt_dt'], date_parser=d)

# Note - for now, it is okay to ignore the warning about mixed types. 

***
## 1. Initial Data Checks (50 points)

First we will take a preliminary look at the data to check that it was loaded correctly and contains the info we need.

The questions to answer at the end of this section:
- Do we have the correct # of columns and rows. 
- Do the records contain data for the questions we want to answer 
- What columns are important? 
- What columns can be dropped?
- What are the data problems?

**1a.** Print the *shape* of the data. Does this match the expectation? (2 points)

In [2]:
# 1a YOUR CODE HERE

print(contrib.shape)

(1125659, 18)


- **1a answer here:** 

The output does match the number of rows and columns of the data as expected

**1b.** Print a list of column names. Are all the columns included that are in the documentation? (2 points)

In [3]:
# 1b YOUR CODE HERE

print(contrib.columns.tolist())

['cmte_id', 'cand_id', 'cand_nm', 'contbr_nm', 'contbr_city', 'contbr_st', 'contbr_zip', 'contbr_employer', 'contbr_occupation', 'contb_receipt_amt', 'contb_receipt_dt', 'receipt_desc', 'memo_cd', 'memo_text', 'form_tp', 'file_num', 'tran_id', 'election_tp']


- **1b answer here:** 

Yes, this does print all the columns in the documentation 

**1c** Print out the first five rows of the dataset. How do the columns `cand_id`, `cand_nm` and `contbr_st` look? (3 points)

In [4]:
# 1c YOUR CODE HERE
#print(contrib.head())
print(contrib[['cand_id', 'cand_nm', 'contbr_st']].head())

     cand_id                  cand_nm contbr_st
0  P00003392  Clinton, Hillary Rodham        CA
1  P00003392  Clinton, Hillary Rodham        CA
2  P00003392  Clinton, Hillary Rodham        CA
3  P60007168         Sanders, Bernard        CA
4  P60007168         Sanders, Bernard        CA


- **1c answer here:** 

cand_id has the unique ID for each candidate (i.e. P00003392)
cand_num has the name of each candidate (i.e. Bernard Sanders)
contbr_st has the state that each candidate is part of (i.e. CA)

**1d.** Print out the values for the column `election_tp`. In your own words, based on the documentation, what information does the `election_tp` variable contain? Do the values in the column match the documentation? (3 points)

In [5]:
# 1d YOUR CODE HERE
print(contrib['election_tp'].unique())

['P2016' 'G2016' nan 'P2020']


- **1d answer here:** 

Eleciton_tp holds the election type/primary general indicator. The values do match with the documentation provided.

**1e.** Print out the datatypes for all of the columns. What are the datatypes for the `contbr_zip`, `contb_receipt_amt`, `contb_receipt_dt`? (5 points)

In [6]:
# 1e YOUR CODE HERE
print(contrib.dtypes)
print("Data type contbr_zip", contrib['contbr_zip'].dtype)
print("Data type contb_receipt_amt", contrib['contb_receipt_amt'].dtype)
print("Data type contb_receipt_dt",contrib['contb_receipt_dt'].dtype)


cmte_id                      object
cand_id                      object
cand_nm                      object
contbr_nm                    object
contbr_city                  object
contbr_st                    object
contbr_zip                   object
contbr_employer              object
contbr_occupation            object
contb_receipt_amt           float64
contb_receipt_dt     datetime64[ns]
receipt_desc                 object
memo_cd                      object
memo_text                    object
form_tp                      object
file_num                      int64
tran_id                      object
election_tp                  object
dtype: object
Data type contbr_zip object
Data type contb_receipt_amt float64
Data type contb_receipt_dt datetime64[ns]


- **1e answer here:**

Data Type contbr_zip: Object/String, 
Data Type contb_receipt_amt: Numeric/Float,
Data Type contb_receipt_dt: Date/datime64[ns]


**1f.** What columns have the most non-nulls?  Would you recommend to drop any columns based on the number of nulls? (5 points)

In [7]:
# 1f YOUR CODE HERE
not_null = contrib.count()
print("Columns with the most non-null values:")
print(not_null.sort_values(ascending=False))


Columns with the most non-null values:
cmte_id              1125659
cand_id              1125659
tran_id              1125659
file_num             1125659
form_tp              1125659
contb_receipt_dt     1125659
contb_receipt_amt    1125659
contbr_st            1125659
contbr_nm            1125659
cand_nm              1125659
contbr_city          1125633
contbr_zip           1125564
election_tp          1124234
contbr_occupation    1115260
contbr_employer       967757
memo_text             501148
memo_cd               144268
receipt_desc           15045
dtype: int64


- **1f answer here:** 

The columns with the most non-nulls are these: 

cmte_id,              
cand_id,             
tran_id,             
file_num,             
form_tp,             
contb_receipt_dt,     
contb_receipt_amt,    
contbr_st,            
contbr_nm,           
cand_nm 


Based on the number of nulls present in receipt_desc, this column could be a candidate for removal 

**1g.** A column we know that we want to use is the cand_nm column.  From the documentation each candidate is a unique candidate id also. Check data quality of `cand_id` column to see if it matches `cand_nm` column. Specifically check to ensure our targetted candidate 'Bernard Sanders' always has the same cand_id throughout. Any issues with `cand_nm` matching `cand_id`? (5 points)

In [8]:
# 1g YOUR CODE HERE

#Check data quality of cand_id and cand_num
unique_value_column = contrib[['cand_id', 'cand_nm']].drop_duplicates()
cand_nm_check = unique_value_column .duplicated(subset='cand_nm').any()
cand_id_check = unique_value_column .duplicated(subset='cand_id').any()
is_cand_id_matching_cand_nm = not cand_nm_check and not cand_id_check



sanders_data = contrib[contrib['cand_nm'] == 'Sanders, Bernard']

cand_id_unique_value = sanders_data['cand_id'].unique()

is_cand_id_unique = sanders_data['cand_id'].nunique() == 1

if is_cand_id_matching_cand_nm:
    print('cand_id column matches the cand_nm column')
else:
    print("cand_id' does not match match cand_nm")
print("Unique id Bernard Sanders:", cand_id_unique_value)
print("Same value for Bernard Sanders always?:", is_cand_id_unique)


cand_id column matches the cand_nm column
Unique id Bernard Sanders: ['P60007168']
Same value for Bernard Sanders always?: True


- **1g answer here:** 

There are no issues with the cand_id and cand_nm columns matching. In addition, there is a unique id for Bernard Sanders and it is always the same value 

**1h.** Another area to check is to make sure all of the records are from California. Check the `contbr_st` column - are there any records outside of California based on `contbr_st`? (5 points)

In [9]:
# 1h YOUR CODE HERE
state_checker = contrib['contbr_st'].nunique() > 1 and 'CA' not in contrib['contbr_st'].unique()
if state_checker:
    print("There are records from states outside of California.")
else:
    print("All records are from California.")

All records are from California.


- **1h answer here:** 

All of the records are from California

**1i.** The next column to check for the analysis is the `tran_id` column. This column could be the primary key so look for duplicates. How many duplicate entries are there? Any pattern for why are there duplicate entries? (5 points)

In [10]:
# 1i YOUR CODE HERE
tran_id_duplicates = contrib['tran_id'].duplicated().sum()
print('Duplicate count:',tran_id_duplicates)


duplicate_values = contrib[contrib.duplicated(subset='tran_id', keep=False)]

# Print the duplicate entries
print("Duplicate entries:")
print(duplicate_values)



Duplicate count: 3454
Duplicate entries:
           cmte_id    cand_id                    cand_nm  \
57       C00575795  P00003392    Clinton, Hillary Rodham   
90       C00575795  P00003392    Clinton, Hillary Rodham   
91       C00575795  P00003392    Clinton, Hillary Rodham   
174      C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
175      C00574624  P60006111  Cruz, Rafael Edward 'Ted'   
...            ...        ...                        ...   
1125589  C00573519  P60005915        Carson, Benjamin S.   
1125596  C00573519  P60005915        Carson, Benjamin S.   
1125597  C00573519  P60005915        Carson, Benjamin S.   
1125598  C00573519  P60005915        Carson, Benjamin S.   
1125614  C00573519  P60005915        Carson, Benjamin S.   

                     contbr_nm    contbr_city contbr_st     contbr_zip  \
57             SANMUKH, BHAKTA      SAN DIMAS        CA 917,733,736.00   
90             ROBERT, MONKMAN        FREMONT        CA 945,365,810.00   
91             D

- **1i answer here:** 

There are 3454 duplicates entries. The pattern/reason on why there are duplicated entries is because there can be a unique candidate with duplicated values of the same transaction 

**1j.** Another column to check is the `contb_receipt_amt` that shows the donation amounts. How many negative donations are included? What do negative donations mean? Please show at least pull a few rows to look at the records with negative donations. Do these records match with the expectation of why a negative donation would happen? (5 points)

In [11]:
# 1j YOUR CODE HERE

negative_donation_count = (contrib['contb_receipt_amt'] < 0).sum()

print("Negative donations:", negative_donation_count)

negative_donations = contrib[contrib['contb_receipt_amt'] < 0]
print("Rows with negative donations:")
print(negative_donations.head())

Negative donations: 11896
Rows with negative donations:
       cmte_id    cand_id                    cand_nm           contbr_nm  \
19   C00574624  P60006111  Cruz, Rafael Edward 'Ted'    JOLLIFF, RICHARD   
23   C00574624  P60006111  Cruz, Rafael Edward 'Ted'    JOLLIFF, RICHARD   
81   C00574624  P60006111  Cruz, Rafael Edward 'Ted'    JOLLIFF, RICHARD   
190  C00574624  P60006111  Cruz, Rafael Edward 'Ted'    NOWELL, DIANA L.   
213  C00574624  P60006111  Cruz, Rafael Edward 'Ted'  LICHTY, ANDREW MR.   

                contbr_city contbr_st     contbr_zip  \
19                    CHICO        CA 959,289,507.00   
23                    CHICO        CA 959,289,507.00   
81                    CHICO        CA 959,289,507.00   
190  RANCHO SANTA MARGARITA        CA 926,884,928.00   
213               SAN DIEGO        CA 921,096,720.00   

                        contbr_employer     contbr_occupation  \
19                        SELF EMPLOYED               RANCHER   
23                  

- **1j answer here:**

Yes, this does match with the expectation of a negative donation 

**1k.** One more column to look at is the date of donation column. Are there any dates outside of the primary period (defined as 1 Jan 2014 to 7 June 2016)? Are the dates well-formatted for our analysis? (5 points)

In [12]:
# 1k YOUR CODE HERE

start_date = '2014-01-01'
end_date = '2016-06-07'

dates_not_primary = contrib[(contrib['contb_receipt_dt'] < start_date) | (contrib['contb_receipt_dt'] > end_date)]
count_dates_not_primary = len(dates_not_primary)

print('Dates not in primary:')
print(dates_not_primary['contb_receipt_dt'])
print("Number of dates outside the primary period:", count_dates_not_primary)

Dates not in primary:
9932      2013-11-05
9994      2013-11-05
14673     2016-06-30
14682     2016-07-20
14697     2016-07-14
             ...    
1123992   2016-08-02
1123993   2016-08-09
1123994   2016-08-31
1123995   2016-08-31
1123996   2016-07-06
Name: contb_receipt_dt, Length: 451372, dtype: datetime64[ns]
Number of dates outside the primary period: 451372


- **1k answer here:**

Yes, the dates are well formatted for our analysis and there are 451372 dates that are not included in the primary period

**1l.** Finally, answer the initial questions in the cells below (5 points)

**1l.1** Do we have the correct # of columns and rows.

- **1l.1 answer here:**

Yes, we do have the correct # of columns and rows

**1l.2** Do the records contain data for the questions we want to answer?

- **1l.2 answer here:**

Yes, all of the records have sufficient data to carry out our analysis

**1l.3** What columns are important?

- **1l.3 answer here:** 

The most important columns: 

cmte_id,cand_idn,cand_nm,contbr_nm,contbr_city,contbr_st,contbr_zip,contb_receipt_amt,contb_receipt_dt

**1l.4** What columns can be dropped?

- **1l.4 answer here:** 

The columns that can be dropped are: 

form_tp, file_num, tran_id, memo_cd, memo_text, receipt_desc

**1l.5** What are the data problems?

- **1l.5 answer here:**

Some of the data problems include: 
1. contbr_zip has an inconsistent across the values and not in the correct zip code format
2. There are several duplicated columns, especially when analzying tran_id
3. There are several values with NaN  


**1l.6** List any assumptions so far:

- **1l.6 answer here:**

One of the assumptions is that the amount is in local currency for our analaysis. It is also assumed that the candidates are affiliated to a specific party. 


***
## 2. Data filtering and data quality fixes (30 points)

Now that we have a basic understanding of the data, let's filter out the records we don't need and fix the data.

**2a.** From the dataset filter out (remove) any election_tp not in the primary election. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [13]:
# 2a YOUR CODE HERE

filter_primary =  contrib[contrib['election_tp'].str.startswith('P', na=False)]
print(filter_primary.shape)

(810488, 18)


**2b.** From the dataset filter out (remove) any candidate that is not Bernie Sanders. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [14]:
# 2b YOUR CODE HERE

filter_not_sanders = contrib[(contrib['cand_nm'] == 'Sanders, Bernard')]
print(filter_not_sanders.shape)

(407172, 18)


**2c.** The `contbr_zip` column is not formatted well for our analysis. Make a new zipcode column that is the five-digit zipcodes. Filter out any records outside of California based on the zipcode. Print/show the shape of the dataframe after the filtering is complete. (10 points).

- You will have to research what the valid 5-digit zipcodes for California are!

In [15]:
# 2c YOUR CODE HERE
contrib['zipcode_clean'] = contrib['contbr_zip'].apply(lambda x: str(x)[:5])

# Step 2: Filter out any records outside of California based on the zip code
california_data = contrib[contrib['zipcode_clean'].str.startswith('9')]

# Step 3: Print/show the shape of the DataFrame after filtering
print(california_data.shape)



(1125423, 19)


**2d.** The receipt amount column has negative donations. After talking with your team, a decision was made that the best course of action is to remove these negative values so that the donation count and amount is more accurate. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [16]:
# 2d YOUR CODE HERE

contrib = contrib[contrib['contb_receipt_amt'] >= 0]
print("After filtering out negative donations shape:",contrib.shape)


After filtering out negative donations shape: (1113763, 19)


**2e.** From the dataset drop any columns that won't be used in the analysis. Print/show the shape of the dataframe after the dropping is complete. What columns did you drop and why? (5 points)

In [17]:
# 2e YOUR CODE HERE

rm_columns = ['form_tp', 'file_num', 'tran_id', 'memo_cd', 'memo_text', 'receipt_desc']

contrib = contrib.drop(columns=rm_columns)

print("Shape after dropping unnecessary columns:",contrib.shape)


Shape after dropping unnecessary columns: (1113763, 13)


- **2e answer here:**

I decided to drop columns form_tp,file_num,tran_id,memo_cd,memo_text and receipt_desc as they are just descriptive columns. They do give valuable information for the user, but in our analysis they are not needed.

**2f.** List any assumptions that you made up to this point:

- **2f answer here:**


One of the assumptions that have been made is that all of the zip codes included in the list are valid. In addition, we assumed that removing negative donors will actually give a more accurate view of the total donation amount. 

***
## 3. Answering the questions (20 points)

Now that the data is cleaned and filterd - let's answer the two questions from your boss!

**3a.** Which zipcode had the highest count of contributions and the most dollar amount? (10 points)

In [18]:
# 3a YOUR CODE HERE
zipcode_contributions = contrib.groupby('zipcode_clean')['contb_receipt_amt'].agg(['count', 'sum'])
count_zipcode = zipcode_contributions['count'].sort_values(ascending=False).nlargest(1).index[0]
dollar_zipcode = zipcode_contributions['sum'].sort_values(ascending=False).nlargest(1).index[0]


print("Zipcode with the highest count of contributions:", count_zipcode)
print("Zipcode with the most dollar amount:", dollar_zipcode)

Zipcode with the highest count of contributions: 94110
Zipcode with the most dollar amount: 90049


- **3a answer here:** 

**3b.** What day(s) of the month do most people donate? (10 points)

In [19]:
# 3b YOUR CODE HERE

contrib['day_of_month'] = contrib['contb_receipt_dt'].dt.day
donation_days = contrib['day_of_month'].value_counts()

# Find the day(s) with the highest count of donations
sorted_donation_days = donation_days.sort_values(ascending=False)
highest_donation_day = donation_days.index[0]
second_highest_day = sorted_donation_days.index[1]
third_highest_day = sorted_donation_days.index[2]
# Print the result
print('Highest donation day:',highest_donation_day)
print('Second donation day:',second_highest_day)
print('Third donation day:',third_highest_day)


Highest donation day: 29
Second donation day: 30
Third donation day: 28


- **3b answer here:** 

Highest donation day: 29

Second donation day: 30

Third donation day: 28

## If you have feedback for this homework, please submit it using the link below:

http://goo.gl/forms/74yCiQTf6k